source/h2_stream.c

```diff
@@ -143,3 +303,158 @@ int aws_h2_stream_on_activated(struct aws_h2_stream *stream, bool *out_has_outgo
+            AWS_H2_STREAM_LOGF(
+                ERROR, stream, "Incoming header callback raised error, %s", aws_error_name(aws_last_error()));
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+
+malformed:
+    return s_send_rst_and_close_stream(stream, AWS_ERROR_HTTP_PROTOCOL_ERROR);
+}
+
+int aws_h2_stream_on_decoder_headers_end(
+    struct aws_h2_stream *stream,
+    bool malformed,
+    enum aws_http_header_block block_type) {
+
+    AWS_PRECONDITION_ON_CHANNEL_THREAD(stream);
+
+    /* Not calling s_check_state_allows_frame_type() here because we already checked
+     * at start of HEADERS frame in aws_h2_stream_on_decoder_headers_begin() */
+
+    if (malformed) {
+        AWS_H2_STREAM_LOG(ERROR, stream, "Headers are malformed");
+        return s_send_rst_and_close_stream(stream, AWS_ERROR_HTTP_PROTOCOL_ERROR);
+    }
+
+    switch (block_type) {
+        case AWS_HTTP_HEADER_BLOCK_INFORMATIONAL:
+            AWS_H2_STREAM_LOG(TRACE, stream, "Informational 1xx header-block done.");
+            break;
+        case AWS_HTTP_HEADER_BLOCK_MAIN:
+            AWS_H2_STREAM_LOG(TRACE, stream, "Main header-block done.");
+            stream->thread_data.received_main_headers = true;
+            break;
+        case AWS_HTTP_HEADER_BLOCK_TRAILING:
+            AWS_H2_STREAM_LOG(TRACE, stream, "Trailing 1xx header-block done.");
+            break;
+        default:
+            AWS_ASSERT(0);
+    }
+
+    if (stream->base.on_incoming_header_block_done) {
+        if (stream->base.on_incoming_header_block_done(&stream->base, block_type, stream->base.user_data)) {
+            AWS_H2_STREAM_LOGF(
+                ERROR,
+                stream,
+                "Incoming-header-block-done callback raised error, %s",
+                aws_error_name(aws_last_error()));
+            return AWS_OP_ERR;
+        }
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+int aws_h2_stream_on_decoder_end_stream(struct aws_h2_stream *stream) {
+    AWS_PRECONDITION_ON_CHANNEL_THREAD(stream);
+
+    /* Not calling s_check_state_allows_frame_type() here because END_STREAM isn't
+     * an actual frame type. It's a flag on DATA or HEADERS frames, and we
+     * already checked the legality of those frames in their respective callbacks. */
+
+    if (stream->thread_data.state == AWS_H2_STREAM_STATE_HALF_CLOSED_LOCAL) {
+        /* Both sides have sent END_STREAM */
+        stream->thread_data.state = AWS_H2_STREAM_STATE_CLOSED;
+        AWS_H2_STREAM_LOG(TRACE, stream, "Received END_STREAM. State -> CLOSED");
+
+        /* Tell connection that stream is now closed */
+        if (aws_h2_connection_on_stream_closed(
+                s_get_h2_connection(stream),
+                stream,
+                AWS_H2_STREAM_CLOSED_WHEN_BOTH_SIDES_END_STREAM,
+                AWS_ERROR_SUCCESS)) {
+            return AWS_OP_ERR;
+        }
+
+    } else {
+        /* Else can't close until our side sends END_STREAM */
```

Comment:
I was going to say why we don't check if the state is valid for receiving an "end stream". Then, I found out it will only be here if the stream is in the active_stream_map.